### PR TITLE
Store onboarding > add product action handling

### PR DESCRIPTION
**Pull request by jaclync**
Part of #8908 

## Description

⚠️ This PR is likely having merge conflicts with https://github.com/woocommerce/woocommerce-ios/pull/9033 that will be merged before this PR

This PR navigated the user to the Add Product flow after they tap on the `Add your first product` task. Notes on some related changes:

- Moved `StoreOnboardingCoordinator` from where the onboarding card collapsed state is presented (`DashboardViewController`) to within the hosting controller `StoreOnboardingViewHostingController`. The coordinator handles all navigation actions from the view `StoreOnboardingView`, including `View all` (showing expanded state which instantiates a new `StoreOnboardingViewHostingController` with its own coordinator internally), and all the tasks. The current implementation shares the same coordinator and there's a bug where the actions don't work in the expanded state because the shared coordinator is using the incorrect navigation controller (dashboard)
- In `AddProductCoordinator`, `sourceView` is updated to be optional to allow `nil` value since we can't pass the tapped task view as a `UIView` from SwiftUI - lemme know if you know any workarounds. This change affects the tablet presentation of the bottom sheet, but I think we should migrate to the iOS 15+ bottom sheet (should be feasible from my HACK Week attempt for the inventory scanner)
- In `StoreOnboardingCoordinator`, it handles the task tap action by starting `AddProductCoordinator`

## Testing instructions

- Log in if needed
- In the My store tab, tap `Add your first product` --> a bottom sheet should be shown to create a product
- Continue to publish a new product --> all the navigation flow should work smoothly

## Screenshots

https://user-images.githubusercontent.com/1945542/224641179-0f6e1cae-fdc5-42ab-a287-2a6c448839cb.mov

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Review by selanthiraiyan — COMMENTED** [review]
I cannot find the onboarding list in the "My Store" tab. 🤔

**Comment by jaclync** [comment]
> I cannot find the onboarding list in the "My Store" tab. 🤔

That's odd. Did you try scrolling to the bottom of the stats, rotating back and forth, or using the view debugger to check the UI hierarchy? 🤔 

I just merged the latest `trunk` after resolving some merge conflicts, so that it has the latest changes. Can you try again when you get a chance?

**Comment by selanthiraiyan** [comment]
I pulled the latest changes, deleted the app from the simulator and installed a fresh build, and still, I don't see the onboarding list. 

https://user-images.githubusercontent.com/524475/224886681-9c286d7f-73cf-46cf-b507-75e1da6379a2.mp4

I don't see it on the view debugger either.
Screenshot 2023-03-14 at 9 00 21 AM

Tested in,

- iPhone 14 simulator running iOS 16.2. 
- iPhone 13 Pro Max running iOS 16.1.1

I don't see the onboarding list on both simulator and device. I am not sure what I am missing here.

**Comment by jaclync** [comment]
> I don't see the onboarding list on both simulator and device. I am not sure what I am missing here.

@selanthiraiyan have you tried switching to another store?

**Comment by jaclync** [comment]
After taking a look at the changes, it's possible that the `Site` isn't available when `DashboardViewController.showOnboardingCard` is called. I'm guessing you were landing on the My store after logging in. I'll look into removing the `Site` dependency and update here.

**Comment by jaclync** [comment]
@selanthiraiyan 👋 I just updated the PR in https://github.com/woocommerce/woocommerce-ios/pull/9110/commits/b964564f80abea3f99589ebc5f4b19686331ad15 to also observe the site so that a site is always available when setting up the onboarding card. `Site` is required for custom domain action handling. I also tested logging out and in, and was able to see the onboarding card. Lemme know if this issue still persists!

File: WooCommerce/Classes/ViewRelated/Dashboard/Onboarding/StoreOnboardingCoordinator.swift [hunk 1]
```diff
@@ -41,6 +46,16 @@ final class StoreOnboardingCoordinator: Coordinator {
 }
 
 private extension StoreOnboardingCoordinator {
+    @MainActor
+    func addProduct() {
+        let coordinator = AddProductCoordinator(siteID: site.siteID, sourceView: nil, sourceNavigationController: navigationController)
+        self.addProductCoordinator = coordinator
+        coordinator.onProductCreated = { _ in
+            #warning("Analytics when a product is added from the onboarding task")
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/ViewRelated/Dashboard/Onboarding/StoreOnboardingCoordinator.swift:54]: 👍

**Review by selanthiraiyan — APPROVED** [review]
Thank you for the updates, Jaclyn!

This works as expected. 🚢

**Pull request merged**